src/theory/uf/theory_uf_rewriter.h

```diff
@@ -104,6 +126,49 @@ class TheoryUfRewriter {
   static inline void init() {}
   static inline void shutdown() {}
 
+public: //conversion between HO_APPLY AND APPLY_UF
+  // converts an APPLY_UF to a curried HO_APPLY e.g. (f a b) becomes (@ (@ f a) b)
+  static Node getHoApplyForApplyUf(TNode n) {
+    Assert( n.getKind()==kind::APPLY_UF );
+    Node curr = n.getOperator();
+    for( unsigned i=0; i<n.getNumChildren(); i++ ){
+      curr = NodeManager::currentNM()->mkNode( kind::HO_APPLY, curr, n[i] );     
+    }
+    return curr;
+  }
+  // converts a curried HO_APPLY into an APPLY_UF e.g. (@ (@ f a) b) becomes (f a b)
+  static Node getApplyUfForHoApply(TNode n) {
+    Assert( n.getType().getNumChildren()==2 );
+    std::vector< TNode > children;
+    TNode curr = decomposeHoApply( n, children, true );
+    // if operator is standard
+    if( isStdApplyUfOperator( curr ) ){
+      return NodeManager::currentNM()->mkNode( kind::APPLY_UF, children );
+    }
+    // cannot construct APPLY_UF if operator is partially applied or is not standard       
+    return Node::null();
+  }
+  // gets arguments, returns operator of a curried HO_APPLY node
+  static Node decomposeHoApply(TNode n, std::vector<TNode>& args, bool opInArgs = false) {
+    TNode curr = n;
+    while( curr.getKind() == kind::HO_APPLY ){
+      args.push_back( curr[1] );
+      curr = curr[0];        
+    }
+    if( opInArgs ){
+      args.push_back( curr );
+    }
+    std::reverse( args.begin(), args.end() );
+    return curr;
+  }
+  /** returns true if this node can be used as an operator of an APPLY_UF node
+   * f: Int -> Int, g : Int -> Int
+   * forall x : ( Int -> Int ), y : Int. (x y) = (f 0)
+   * Then, f and g are standard APPLY_UF operators, but (ite C f g), (lambda x1. (f x1)) as well as variable x above are not.  
+   */
+  static inline bool isStdApplyUfOperator(TNode n){
+    return n.isVar() && n.getKind()!=kind::BOUND_VARIABLE;
```

Comment:
Thanks for the clarifying comment.  I still like "isUninterpretedFunctionSymbol" better - f and g are uninterpreted function symbols, while the other examples you gave are not.  Also, I think it should check the type (the current function could return true for an integer variable for example).  Finally, I think it would belong better as a static method of the TheoryUF class instead of the rewriter.

Reply:
The reason I don't type check is for efficiency. One should be allowed to call the function with no overhead if we already know that it has a function type (profiling cvc4 in the past has indicated that getType can be up to ~5% of total time in some cases). So isUninterpretedFunction could be a separate function that also checked its type. However, I don't have a need for this function at the moment.

I don't think it belongs in the TheoryUF class, since its use isn't specific to the decision procedure for UF.  I think the options are to either leave it where it is or move it to Node (e.g. where isVar is), where the latter is best if these are generally necessary functions.  However in the ho-smt branch, I managed to not have to refer to these functions outside of the UF theory.

Reply:
I know this seems like I'm being picky, but even if it's not used elsewhere, this is a very general-purpose function and so it doesn't belong here as-is.  How about we move it to Node and call it isFreeVar?

Reply:
Although its body is identical to a general purpose function, this isn't a general purpose function.  

The intention of this function is to say whether a node can/should be considered as an operator of an APPLY_UF node. At the moment, this coincides with "isFreeVar", but this may conceivably change to "isVar", or conversely, something more liberal in the future, depending on the approach we find works best in practice (there are still decisions about when to convert to and from an applicative encoding in an HO SMT approach that we haven't fully worked out in empirical testing).  I need for this function to change uniformly in all occurrences with future updates but only in occurrences where its meaning is what I mention here. Otherwise I would need to distinguish "should this call to isFreeVar be updated now to reflect my new thoughts of which nodes can be considered as operators of APPLY_UF?".